### ISSUE TITLE
Session resume terminology

### ISSUE THREAD

**Issue by kroeckx** [tlswg/tls13-spec#847]
There seem to be 3 cases left where we the draft talks about resumption:
- In 4.5.1: Any ticket MUST only be resumed with a cipher suite that has the same KDF hash [...]
- In 6.1: Failure to properly close a connection does not prohibit a session from being resumed.
- In 1.2: Restate rule from RFC 6066 that you can't resume unless SNI is the  same. (Talking about 4.5.1)

I think the term resume is misleading, you aren't resuming anything. You're really reusing parameters from that session, and can have multiple connections to the same server that use the same parameters.

**Comment by jensguballa** [comment]
The term "resumption" occurs 38 times in the draft. :-)

Anyway, I am more interested in a good definition of the term "session". 

Some thoughts:
  - the term "session" is only required for resumption. That's my observation: if resumption would not be a feature of TLS, we would not need the term "session" at all.
  - a handshake produces a context, which comprises:
    - resumption_master_secret
    - the used KDF hash function
    - the fact that the server is authenticated
    - an indication if the client is authenticated or not
    - SNI (if provided by the client)
    - others (e.g. protocol version, cipher method, etc)
  - this context is used when generating a NewSessionTicket, or we can say: a ticket refers to this context
  - We can give this context a name: "session"
  - Thus each handshake produces a new unique session (because the resumption_master_secret will be unique)
  - A handshake may bootstrap from a previous session (via a ticket, "session resumption"), but even in this case a new session will be created
  - There is a one-to-one relationship between a connection and a session (the one created)
  - multiple connections may bootstrap from the same session

Do these thoughts make sense? If so, then probably some more considerations are needed, e.g. regarding externally provided PSKs ("Are these external sessions? Or external tickets?").

**Comment by kaduk** [comment]
Many of those thoughts make sense, yes.

In some sense we may think about TLS connections more than TLS sessions, with a connection being defined as starting at initial ClientHello and ending at TCP close, timeout, or termination by alert.  A TLS session is then a way to extend some context from one connection to another (which causes trouble for the generic "one-to-one releationship between connection and session" you list above).

In some sense this is all editorial, though, so we may be subject to @ekr 's discretion.

**Comment by katrielalex** [comment]
(moving discussion from #867) @jensguballa points out that the term "session key" appears throughout the appendix, starting with 
>  A "session key" (the master secret) from which can be derived a set of working keys.

Sessions and session keys are terminology from academic security models (an authenticated key exchange protocol is defined to output a session key) which aren't really used elsewhere in the RFC. I suggest one of two options:
 1. replace "session key" by "master secret" and "session" by "context" (or something) throughout the document
 1. add a bit more explanation at the top of §Handshake to comment on the choice of terminology and the definitions of sessions and session keys

BTW I think a session should be defined locally (a client has a session which outputs a session key), and the security property says roughly that if a client thinks its session is with some server S then it really is with server S. Happy to PR for the latter option.

**Comment by cascremers** [comment]
Both options seem reasonable, but I would prefer to avoid introducing terminology that is not really needed, leading to option 1 above: replacing "session key" and "session" by other terms that are used more widely in the spec.

(If "session" is used somewhere in a non-local meaning, we need to carefully consider what to replace it with. Everything else is probably straightforward.)

**Comment by jensguballa** [comment]
I definitely support option 1. Maybe, "session" can be replaced by "PSK context". NewSessionTicket would become NewPSKContextTicket, which would make quite a lot of sense, IMO.

**Comment by jensguballa** [comment]
If no more feedback is received, does that mean that everyone agrees silently and that a PR is expected? ;-)

**Comment by ekr** [comment]
I think we need to continue with the current use of TLS session, confusing
though it is.

Anything else will be even more confusing.

**Comment by richsalz** [comment]
Strongly agree.

**Comment by katrielalex** [comment]
In that case, I have more questions :) At a high level I support @jensguballa's point from above that the term "session" is confusing: sometimes it means a connection, and sometimes it means a nebulous set of connections.

- is "session state" the same as connection state, or a subset, or a superset?
- are "tickets" and "session tickets" the same? if not, should the language be made consistent, as both are used in the doc?
- what exactly is "resuming a session"---performing a PSK resumption with a ticket that comes from a previous connection? If a client reuses a ticket, resulting in two connections, are those connections both the same session?
- does every session have a unique session key? (if yes, you can't have session key = master secret)
- the PSK binder says that it binds two sessions together, but the rest of the doc says that you resume the first session. which is it?

**Comment by jensguballa** [comment]
IMO these are valid questions caused by a lack of precise and consistent terminology . Would like to add one:

- when does a session start to exist? With sending/receiving the ClientHello (no master secret [= session id] yet available)? Or with sending/receiving the ServerHello?

I think the term "session" might have been appropriate for TLS1.0-1.2 in some way, as there the identical master secret is used by two different connections, and thus the term "session resumption" made some sense. However, with TLS1.3 a dedicated resumption master secret is used for the NewSessionTicket message. Thus the resumption mechanism has been changed drastically, but still the TLS1.0-1.2 terminology "session resumption" is used. Hm.

So honestly, I don't get the point why sticking to "session" should be less confusing than removing it. Different mechanisms in TLS1.3 justify different (better!) terminology.

The current version of the draft is using terms like "resumption via PSK", "1-RTT handshake with PSK resumption", "1-RTT handshake with PSK resumption", "This extension also restricts the modes for use with PSK resumption". So not using the term "session" here works quite well. :-)

**Comment by ekr** [comment]
OK, here's what I'm going to do: A lot of PR #895 is just good cleanup regardless of the resolution of this question. I'm going to take a subset of that and then revisit the question once I can read the whole draft in that context.

**Comment by cascremers** [comment]
Side note: unless I'm mistaken, it seems that "resumption" is conceptually more similar to a "fork" or "branch", since the resulting connection can exist in parallel to the original connection.

**Comment by ekr** [comment]
@cascremers: yes, that's correct

**Comment by ekr** [comment]
@katrielalex @jensguballa: I took most of PR#895, so now the only use of "session" (except in PDU names and in references to previous versions) is in the security analysis section. I kept the term there because it's the one that appears on [CK01] and so SK-security is confusing without it. If people feel strongly, I could imagine pulling in the rest of PR#895 that made those changes, but I think that's a separate PR so closing this issue.

**Comment by davidben** [comment]
Actually, is the "Uniqueness of the session key" claim in E.1 even true? We're saying the master secret is the session key, but the base secrets in the key schedule don't incorporate the handshake transcript. That gets mixed in at the Derive-Secret branches.

Notably, if I do psk_ke resumption of the same PSK twice, I'll have the same master secret both times.

**Comment by ekr** [comment]
Yah, you're right. Some sort of mental race condition where I was writing as if it were the old key schedule :)  The right key here is the various derived key Thanks for catching that. Fixed.

**Comment by ekr** [comment]
https://github.com/tlswg/tls13-spec/commit/3996e459d1bc697b0b23a048cbe94496140bc3b5
